go/common/interrupt/interrupt.go

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2024 Fantom Foundation
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at fantom.foundation/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+package interrupt
+
+import (
+	"context"
+	"log"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/Fantom-foundation/Carmen/go/common"
+)
+
+const ErrCanceled = common.ConstError("export was interrupted")
+
+// IsContextDone returns true if the given context CancelFunc has been called.
+// Otherwise, returns false.
+func IsContextDone(ctx context.Context) bool {
+	select {
+	case <-ctx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+// Catch catches SIGTERM and SIGINT signals and
+// prevents export from corrupting database by canceling its context.
+func Catch(parent context.Context) context.Context {
+	ctx, cancel := context.WithCancel(parent)
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM, syscall.SIGINT)
+	go func() {
+		defer signal.Stop(c)
+		select {
+		case <-c:
+			log.Println("closing, please wait until proper shutdown to prevent database corruption")
```

Comment:
I think we should notify the user that the kill signal has been registered by the app. How about `closing, please wait until proper shutdown` to make it more generic?

Reply:
My point was that now when this functionality is generally available here, it can be registered and triggered from everywhere. Hence, there may be situations where telling the user to wait does not make sense. Think of scenarios where carmen is embedded in a third party client and interrupt comes from another component. There is no way for user to decide if he waits or not. 

The proper solution would be to propagate this signal to registered observers and they decide what to do, but since we do not need it at the moment, I would not complicate this PR too much. 

I would suggest at least to make the log as generic as possible, perhaps just printing "shutting down".

Since we do not have logging anywhere in Carmen, I asked at the beginning just for removing this log here.  But if you feel it is important to have it, I guess there is no harm to log.

Reply:
I understand, but if the client needs to use this feature it prevents some ungraceful closing hence telling user to wait until its properly closed should be there. `Shutting down` is the most generic one - lets use it.

Reply:
Or atleast some kind of log. I can imagine a scenario where used wanted to interrupt the process (even the export) and nothing happens. This would cause confusion. Now that we at least log `shutting down` the user know that something is happening.